src/main/java/uk/gov/hmcts/reform/blobrouter/services/EnvelopeService.java

```diff
@@ -132,7 +135,32 @@ public void saveEvent(UUID envelopeId, EventType eventType) {
     }
 
     @Transactional(readOnly = true)
-    public List<Envelope> getEnvelopes(String blobName, String containerName, LocalDate date) {
-        return envelopeRepository.findEnvelopes(blobName, containerName, date);
+    public List<Tuple2<Envelope, List<EnvelopeEvent>>> getEnvelopes(
+        String blobName,
+        String containerName,
+        LocalDate date
+    ) {
+        List<Envelope> envelopes = envelopeRepository
+            .findEnvelopes(blobName, containerName, date);
+
+        if (envelopes.isEmpty()) {
+            return emptyList();
+        } else {
+            List<UUID> envelopeIds = envelopes.stream().map(e -> e.id).collect(toList());
+
+            List<EnvelopeEvent> envelopeEvents = eventRepository.findForEnvelopes(envelopeIds);
+
+            Map<UUID, List<EnvelopeEvent>> eventsByEnvelopeIds = envelopeEvents
+                .stream()
+                .collect(groupingBy(envelopeEvent -> envelopeEvent.envelopeId));
+
+            return envelopes
+                .stream()
+                .map(envelope -> Tuples.of(
+                    envelope,
+                    eventsByEnvelopeIds.get(envelope.id)
```

Comment:
If, for some reason, there are no events for the given envelope, this will be null. Should we make it an empty list in such case?

Reply:
Updated the service to return the empty list when no events exist for envelope